rosidl_generator_c/cmake/rosidl_generator_c_generate_action_interfaces.cmake

```diff
@@ -0,0 +1,149 @@
+  if(_extension STREQUAL ".action")
+    set(_allowed_parent_folders "action")
+    if(NOT _parent_folder IN_LIST _allowed_parent_folders)
+      message(FATAL_ERROR "Interface file with unknown parent folder: ${_idl_file}")
+    endif()
+  else()
+    message(FATAL_ERROR "Interface file with unknown extension: ${_idl_file}")
+  endif()
+  get_filename_component(_msg_name "${_idl_file}" NAME_WE)
+  string_camel_case_to_lower_case_underscore("${_msg_name}" _header_name)
+  list(APPEND _generated_files
+    "${_output_path}/${_parent_folder}/${_header_name}.h"
+    "${_output_path}/${_parent_folder}/${_header_name}__type_support.h"
+  )
+endforeach()
+
+set(_dependency_files "")
+set(_dependencies "")
+foreach(_pkg_name ${rosidl_generate_action_interfaces_DEPENDENCY_PACKAGE_NAMES})
+  foreach(_idl_file ${${_pkg_name}_INTERFACE_FILES})
+  get_filename_component(_extension "${_idl_file}" EXT)
+  if(_extension STREQUAL ".msg")
+    set(_abs_idl_file "${${_pkg_name}_DIR}/../${_idl_file}")
+    normalize_path(_abs_idl_file "${_abs_idl_file}")
+    list(APPEND _dependency_files "${_abs_idl_file}")
+    list(APPEND _dependencies "${_pkg_name}:${_abs_idl_file}")
+  endif()
+  endforeach()
+endforeach()
+
+set(target_dependencies
+  "${rosidl_generator_c_BIN}"
+  ${rosidl_generator_c_GENERATOR_FILES}
+  "${rosidl_generator_c_TEMPLATE_DIR}/action.h.em"
+  "${rosidl_generator_c_TEMPLATE_DIR}/action__type_support.h.em"
+  ${rosidl_generate_action_interfaces_IDL_FILES}
+  ${_dependency_files})
+foreach(dep ${target_dependencies})
+  if(NOT EXISTS "${dep}")
+    get_property(is_generated SOURCE "${dep}" PROPERTY GENERATED)
+    if(NOT ${_is_generated})
+      message(FATAL_ERROR "Target dependency '${dep}' does not exist")
+    endif()
+  endif()
+endforeach()
+
+set(generator_arguments_file "${CMAKE_CURRENT_BINARY_DIR}/rosidl_generator_c__generate_action_interfaces__arguments.json")
+rosidl_write_generator_arguments(
+  "${generator_arguments_file}"
+  PACKAGE_NAME "${PROJECT_NAME}"
+  ROS_INTERFACE_FILES "${rosidl_generate_action_interfaces_IDL_FILES}"
+  ROS_INTERFACE_DEPENDENCIES "${_dependencies}"
+  OUTPUT_DIR "${_output_path}"
+  TEMPLATE_DIR "${rosidl_generator_c_TEMPLATE_DIR}"
+  TARGET_DEPENDENCIES ${target_dependencies}
+)
+
+add_custom_command(
+  OUTPUT ${_generated_files}
+  COMMAND ${PYTHON_EXECUTABLE} ${rosidl_generator_c_BIN}
+  --generator-arguments-file "${generator_arguments_file}"
+  DEPENDS ${target_dependencies}
+  COMMENT "Generating C++ type support dispatch for ROS interfaces"
+  VERBATIM
+)
+
+# generate header to switch between export and import for a specific package
+set(_visibility_control_file
+  "${_output_path}/action/rosidl_generator_c__visibility_control.h")
+string(TOUPPER "${PROJECT_NAME}" PROJECT_NAME_UPPER)
+configure_file(
+  "${rosidl_generator_c_TEMPLATE_DIR}/rosidl_generator_c__action_visibility_control.h.in"
+  "${_visibility_control_file}"
+  @ONLY
+)
```

Comment:
@apojomovsky @sloretz seems we're not installing the generated `_visibility_control_file`.

Reply:
Installed visibility file in db68926c78de370b35d322c25197f551fbc7a21a